### PR TITLE
Implement nesting in cover image block

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
This PR implements the concepts described in https://github.com/WordPress/gutenberg/issues/5448.
E.g: 
```
<InnerBlocks
	template={ [
		[ 'core/paragraph', {
			align: 'center',
			fontSize: 37,
			placeholder: 'Write title…',
			textColor: '#fff',
		} ],
	] }
	allowedBlockNames={ [ 'core/button', 'core/heading', 'core/paragraph', 'core/subhead' ] }
/>
```

As future work template_lock will be added. This opens the path for a whole new type of blocks.
Blocks that make use of nesting and just apply a special arrangement/design to existing blocks. 
E.g: we can have a package of landing page blocks, where all of the blocks are based on paragraphs, headings, buttons etc... arranging them in a special way and applying some styles. Themes may take advantage of this to create theme-specific blocks.

The fact that the InnerBlocks settings are just props has some advantages e.g: depending on the settings in the block inspector we may allow different blocks. But it is much more complex to implement when compared to `Add a property to the block API, allowedRootBlockNames: [ ... ] `.

Right now this implementation is not very clean ( being nice to myself ) we are using a property in the state the save the current InnerBlock settings of each block uid with custom settings. This was required because we need to globally know the allowed blocks, so the inserter can filter them. Maybe there is a better way to solve this problem. I'm thinking about an impossible way to improve this and I'm totally open to suggestions.

To define templates we are using the same logic as the one used in CPT's but this makes the data structure strange in javascript (until know templates were only defined in PHP so this was unnoticeable).

Cover image block was updated to make use of this concepts.

## Know problems / related work

**Using the autocomplete, we can insert allowed blocks.** Autocomplete logic is different from other inserters and we already have some problems e.g: if the template lock the blocks, autocomplete inserter is not aware of the lock and then we get a big js error because onReplace does not exist in this case. In parallel, I will improve the logic of autocomplete inserter so changes required here are easier.
 

## Testing
Test the cover image block try to add blocks remove see things work as expected. The design will benefit from https://github.com/WordPress/gutenberg/pull/5658.
Test that templates also work with layouts by adding this template to columns block.
```
                    template={ [
                        [ 'core/paragraph', {
                            placeholder: 'Enter column 1 content…',
                            layout: 'column-1',
                        } ],
                        [ 'core/paragraph', {
                            placeholder: 'Enter column 2 content…',
                            layout: 'column-2',
                        } ],
                    ] }
```

Test that we are backcompatible with existing cover images by pasting the following code in the code editor and verifying it works as expected:
```
<!-- wp:cover-image {"url":"http://via.placeholder.com/350x150","id":10559} -->
<div class="wp-block-cover-image has-background-dim" style="background-image:url(http://via.placeholder.com/350x150)">
	<p class="wp-block-cover-image-text">Test</p>
</div>
<!-- /wp:cover-image -->
```

## Screenshots (jpeg or gifs if applicable):
image

### CODE REVIEW

**Comment by aduth** [comment]
Related: #4097 (as far as prior art in restricting block types shown in the inserter)

**Comment by phpbits** [comment]
@jorgefilipecosta  looking forward to this :) Thanks!

File: editor/store/reducer.js
Comment (youknowriad) [inline editor/store/reducer.js]: Can you add a docBlock to explain what's this reducer for?

File: editor/components/block-list/block.js
Comment (youknowriad) [inline editor/components/block-list/block.js]: I think synchronizing the template is not needed here because we only synchronize templates if the template is locked but we don't support locking yet for inner Blocks.

Unless I'm missing something?

File: editor/utils/block-list.js
Comment (youknowriad) [inline editor/utils/block-list.js]: As far As I understand we have the `uid` at this point, so we can call the action directly in this component to avoid the need to add a new callback to this HoC?

File: editor/utils/block-list.js
Comment (youknowriad) [inline editor/utils/block-list.js]: Same for this callback: can w call the action directly in this component to avoid the need to add a new callback to this HoC?

File: editor/components/inserter-with-shortcuts/index.js
Comment (youknowriad) [inline editor/components/inserter-with-shortcuts/index.js]: Maybe we could move this logic to a selector to test it properly:

`getSupportBlocks( state, uid, enabledBlockTypes )`

File: editor/components/inserter/index.js
Comment (youknowriad) [inline editor/components/inserter/index.js]: It looks like this prop is not used in the current component, can we move to the menu component instead?

Also might be good to factorize, using the selector proposed above?

**Review by youknowriad — COMMENTED** [review]
Nice work here 👍

File: blocks/library/cover-image/index.js
Comment (gziolo) [reply blocks/library/cover-image/index.js]: Should we move it to the settings object in `registerBlockType` to make it extensible? :)

Comment (gziolo) [reply blocks/library/cover-image/index.js]: I meant both `template` and `allowedBlockNames`.

Comment (youknowriad) [reply blocks/library/cover-image/index.js]: As noted in the description, this allows it to be customizable using toolbar/inspector controls.

But this makes me think of something. We should consider the `template` prop here as a `defaultTemplate` because we already have support for nested template. by setting a CPT template with a nested template for the cover image for instance. I think we should do something here where the template prop is used (override it by the global template if available or something)

**Comment by gziolo** [comment]
> To define templates we are using the same logic as the one used in CPT's but this makes the data structure strange in javascript (until know templates were only defined in PHP so this was unnoticeable).

It doesn't look that bad, it is similar to what is used in config files for libraries like Babel. See: https://github.com/WordPress/packages/blob/master/packages/babel-preset-default/index.js#L6-L11

File: blocks/library/cover-image/index.js
Comment (gziolo) [inline blocks/library/cover-image/index.js]: I think `align` should stay in here. Otherwise, it won't be possible to change the alignment of the container.

File: blocks/library/cover-image/index.js
Comment (gziolo) [inline blocks/library/cover-image/index.js]: The same applies in here, unless it is moved to the InnerBlock. In that case `validAlignments` and `getEditWrapperProps` might need to be removed.

File: blocks/library/cover-image/index.js
Comment (gziolo) [reply blocks/library/cover-image/index.js]: I’m missing something here, how can I override those values for Cover Image block? :smile:
I couldn't digest this from this PR.

Comment (mtias) [reply blocks/library/cover-image/index.js]: Can we switch this API to be `allowedBlocks`?

File: blocks/library/cover-image/editor.scss
Comment (jorgefilipecosta) [reply blocks/library/cover-image/editor.scss]: Hi @youknowriad,
The reason why we set a white background is related to the contrast verification. If the background was never set getComputedStyles returns back background (if I remember correctly). So adding a background fixes the problem. If themes change the background it also works well because as long as a background is set we are able to retrieve it. The problem happens when no background is set. As we get back background color when background is not set we cannot differentiate and assume is white background because in fact it may be set to black. I will try to research this in more detail maybe there is a way to detect when the background is not set and in that case assume white.
Replicate from comment https://github.com/WordPress/gutenberg/pull/5273#discussion_r176880358.

A refactoring will happen to contrast checker in order to not rely on this rule. So it will be removed for now.

File: editor/components/block-list/block.js
Comment (jorgefilipecosta) [reply editor/components/block-list/block.js]: Yes synchronizing the template is not needed. But I want to given a template parse and create the block structure it represents turns out that's exactly what `synchronizeBlocksWithTemplate( [], template )` does so I did not wanted to duplicate code and used it :)

File: editor/components/inserter/index.js
Comment (jorgefilipecosta) [reply editor/components/inserter/index.js]: Hi @youknowriad, the prop supported block types is used to compute hasSupportedBlocks flag that's why it was computed in this component.
Regarding the factorize to selector, I will try to make this more generic maybe it will involve selector plus HoC that accesses the context.

File: blocks/library/cover-image/index.js
Comment (jorgefilipecosta) [reply blocks/library/cover-image/index.js]: Yes, the idea I have in mind is that it is no possible to change the alignment of the container. The container takes 100% width. So for example, if we want to right align text we apply the alignment on the nested paragraph block.

File: editor/components/inserter-with-shortcuts/index.js
Comment (gziolo) [reply editor/components/inserter-with-shortcuts/index.js]: Second that 👍

File: blocks/test/fixtures/core__cover-image.html
Comment (gziolo) [inline blocks/test/fixtures/core__cover-image.html]: has-undefined-content is here because contenAlign attribute was removed, but not its corresponding class name.

File: blocks/test/fixtures/core__cover-image.json
Comment (gziolo) [inline blocks/test/fixtures/core__cover-image.json]: It might be nice to keep non-default dim to see the class name included in HTML.

File: blocks/test/fixtures/core__cover-image.html
Comment (gziolo) [inline blocks/test/fixtures/core__cover-image.html]: Is it expected that fot size is not printed as style?

**Review by gziolo — COMMENTED** [review]
I left a few comments to check.

File: blocks/library/cover-image/index.js
Comment (jorgefilipecosta) [reply blocks/library/cover-image/index.js]: Hi @youknowriad,

> But this makes me think of something. We should consider the template prop here as a defaultTemplate because we already have support for nested template. by setting a CPT template with a nested template for the cover image for instance. I think we should do something here where the template prop is used (override it by the global template if available or something).

When we have a CPT template and we set content for a nested block, ( e.g: we set a cover image with a header and a paragraph), during the post-load the cover image is created and the blocks are added inside it. So when our inner logic executes the cover image block already has content, so the template will not be used. It will be the same as when we load a post with a cover image that already has blocks inside it. The template only applies to empty blocks.
So both approaches can work together without problems if a global template exists it already overwrites this setting :)

**Comment by jorgefilipecosta** [comment]
Hi @gziolo and @youknowriad thank you a lot for the feedback it was applied. 
These changes need additional test cases which I will add in parallel. But the logic is ready for a new look.
Regarding extensibility, and allowing plugins to change the inner block settings, unfortunately, this PR does not address it (in order to not increase the size), but my plan is to creates a new filter that allows plugins to change all the props passed to InnerBlocks (allowedBlocks, template, etc...), the filter receives an object with the settings and can return a new one that changes them. There are other ways like a more granular allowedBlocks filter. after we land the core changes (or agree that they are the path to go) I will add a PR that allows plugins to change these settings :)

File: editor/components/inserter-with-shortcuts/index.js
Comment (youknowriad) [inline editor/components/inserter-with-shortcuts/index.js]: There's a tab instead of a space here

File: editor/store/reducer.js
Comment (youknowriad) [inline editor/store/reducer.js]: typo `blocksn`

File: editor/utils/block-list.js
Comment (youknowriad) [inline editor/utils/block-list.js]: Minor: Would it be clearer if we move the HoC after the class declaration? and move it before the array assignment?

```
const MyClass = class Something {};
const MyClassContainer =compose(...)(myClass);
INNER_BLOCK_LIST_CACHE[ uid ] = [
  MyClassContainer, 0
]
```

**Review by youknowriad — COMMENTED** [review]
Nice Work and Thanks for the update, This is great code wise. The remaining piece for me is the deprecated block.

**Comment by jorgefilipecosta** [comment]
Hi @youknowriad, the remaining feedback was addressed. Regarding deprecation, it is handled on separate PR https://github.com/WordPress/gutenberg/pull/5932 as deprecating from attributes to innerBlocks required changes to the API itself.

**Comment by jorgefilipecosta** [comment]
This PR was rebased, deprecation was handled in https://github.com/WordPress/gutenberg/pull/5932 and it was already approved, I think this one should be ready.
@karmatosed would it be possible to do a design/UX check before the merge :)?

**Comment by mtias** [comment]
image

- Paragraphs should default to light-gray, not just the initial one. `textColor` on the default paragraph should also be removed.
- The arrows need some design work here cc @jasmussen 
- The side inserter, quick block shortcuts, sibling line inserter, etc, should all probably default to white.
- The button seems to have a busted line height, but not sure if it's from this PR or master.

image

- First time you insert the block, the toolbar for cover image is mashed together with toolbar for the paragraph block.

image

- The drag placeholder silhouette should probably have some opacity.

**Comment by mtias** [comment]
@jasmussen one possibility to reduce UI overlap would be to make the movers/ellipsis fall just outside the image boundaries for all widths (except, obviously, for full-width).

image

**Comment by mtias** [comment]
Something else I liked was the simplicity of the inline toolbar aligned at the center with the text. In the above, since we don't clearly see the block box, it makes the toolbar appear disconnected.

image

File: editor/store/reducer.js
Comment (aduth) [inline editor/store/reducer.js]: Do we need to track each block here? Or is it just the context of the currently selected block?

Comment (aduth) [reply editor/store/reducer.js]: What is a "nested nesting" ?

Comment (jorgefilipecosta) [reply editor/store/reducer.js]: Hi @aduth, "nested nesting" was a typo it was supposed to be nested settings.

> Do we need to track each block here? Or is it just the context of the currently selected block?

This is complex, right now, in our use cases only the inserter is using the information for the selected block. But this information should be available for all the blocks and not just the selected one. We may add new settings that affect not just the block that is currently selected.
Another important case is drag&drop, we will need to get the allowed blocks not just for the selected block but for the block we are dragging into so we can check if the dragged block can be nested inside it.

File: editor/store/actions.js
Comment (noisysocks) [inline editor/store/actions.js]: This method is missing a JSDoc comment

File: blocks/library/cover-image/editor.scss
Comment (noisysocks) [inline blocks/library/cover-image/editor.scss]: nit: There's some unnecessary extra whitespace here

File: editor/utils/block-list.js
Comment (noisysocks) [inline editor/utils/block-list.js]: nit: This comment needs an extra level of indentation

File: editor/utils/block-list.js
Comment (noisysocks) [inline editor/utils/block-list.js]: nit: We generally prefer to arrange class methods so that React lifecycle methods are immediately below the constructor

File: editor/utils/block-list.js
Comment (noisysocks) [inline editor/utils/block-list.js]: [`componentWillMount` will be marked as unsafe in React 17.](https://reactjs.org/docs/react-component.html#unsafe_componentwillmount) Let's future-proof this code by using `componentDidMount` instead.

File: editor/utils/block-list.js
Comment (noisysocks) [inline editor/utils/block-list.js]: We shouldn't need to bind these two methods since they aren't being used as a callback anywhere, e.g. they aren't passed as an `on[Event]` prop.

**Review by noisysocks — CHANGES_REQUESTED** [review]
Great work on this! 🥇 

It feels weird to me that we're storing static data (i.e. data that comes from the compiled app bundle and not from user input) in the Redux store, but I can see how that makes things really easy for us.

Could you please add unit tests for the new selectors, reducer, and action?

**Comment by jasmussen** [comment]
Great PR. As Matías notes, this surfaces a bunch of UI issues around the movers and toolbars and such, and these issues are really good to surface, so we can address them. The appender also needs some contrast work, so the buttons are visible. Perhaps we actually need to invert those icon colors when used on a dark background, but this is tricky since it's hard to know the background color of the image. 

I have a few things on my table, plus I'm recovering from a stomach bug so I'm frustratingly slow today, but if you like I can take a stab in this branch at trying to fix some of those issues. But it is unlikely to happen today, but hopefully this week otherwise the next. 

Alternatively you can try and take a stab at this yourself and I'll try and follow up with help as soon as I can. 

The good news is there are lots of improvements happening around this specific UI, it's just all in separate unmerged PRs at this moment. Once it all gets merged in, I think we'll be very close.

**Comment by noisysocks** [comment]
> In parallel, I will improve the logic of autocomplete inserter so changes required here are easier.

Heads up that @brandonpayton is working on changing the autocompleter as well. It should be possible to make the it use `selectors.getInserterItems` like the other inserters.

**Comment by jorgefilipecosta** [comment]
Hi @mtias, the design was improved with your suggestions in mind. Now we give more relevance to the nested blocks like in the columns:
image
The buttons are now white as suggested. And the default text Color is white too without styles being set to the default paragraph. Regarding the problems of toolbar mixing up, and button height I think they were not something exclusive of this PR and with the rebase, it looks like they were solved.


Hi @noisysocks, thank you for the feedback it was applied. Regarding the validation errors, it looks like the old cover image are being migrated correctly. What happens is that our validation mechanism when finds a non-valid block, even if the migration exists, throws the errors to the console. The same errors happen when migrating other blocks. I think when blocks are migrated we should not display this errors, or at least display a message saying they were migrated. We can open a different PR for this as it is related to the block API and not related with the changes were.

**Comment by jasmussen** [comment]
Nice work. Not perfect designwise, and that's on me and separate PRs to fix :D — but a definite improvement in the name of shipping.

**Comment by paulwilde** [comment]
I think applying the white backgrounds on the side UI buttons globally rather than just on the Cover Image block would make more sense.

I've built a website where I've adjusted the entire background of the editor from white, and it required me to tweak the background colour of those buttons in order to look right. I also applied it to the multi-select button which appears (Allowing you to mass delete/duplicate blocks).

**Comment by mtias** [comment]
@paulwilde yes, it sounds like we should play more with transparency by default. Also related: https://github.com/WordPress/gutenberg/pull/6141

**Comment by noisysocks** [comment]
Thanks @jorgefilipecosta. This is looking great. Could you please add unit tests for the new selectors, reducer, and action?

**Comment by jorgefilipecosta** [comment]
Hi @noisysocks, test cases were added to all new artifacts. More design changes were applied (transparencies on drag&drop)

**Comment by lsl** [comment]
I've been building against this to implement a [form builder](https://github.com/lsl/jetpack/commit/4b36cb9458d0429b9c736d932388733495570866) for the jetpack contact form module. 

[Working really well so far.](https://github.com/lsl/jetpack/blob/4b36cb9458d0429b9c736d932388733495570866/modules/contact-form/blocks/form/block.jsx)

I'm curious what it would take to add the override for isPrivate discussed in #5448? That change would let me limit my child blocks to only show up in the inspector when within the form context.

Almost certainly out of scope but I would be keen to hear your thoughts in case I need to implement it: 

How would you go about adding an extra list or a flag to alter allowedBlocks behaviour. What I want is to be able to add the extra form blocks but still maintain the default inspector list.

**Comment by lsl** [comment]
Ah yes, that would work well. Looks like my want for preferential treatment in the inspector was already covered in [the](https://github.com/WordPress/gutenberg/issues/5540#issuecomment-372031314) [comments](https://github.com/WordPress/gutenberg/issues/5540#issuecomment-372312248) too.

**Comment by jorgefilipecosta** [comment]
This PR was rebased and some with a tweak added so we can take advantage of the side UI improvements PR https://github.com/WordPress/gutenberg/pull/6141.

File: editor/store/actions.js
Comment (noisysocks) [inline editor/store/actions.js]: Typo here: should be `UPDATE_BLOCK_LIST_SETTINGS` (without the `D`)

**Review by noisysocks — APPROVED** [review]
This looks good to go. Thanks for adding tests.

The only thing I noticed is some weird z-index behaviour going on when you insert a Button into a Cover Image:

screen shot 2018-04-19 at 14 16 36

**Comment by jorgefilipecosta** [comment]
Hi @noisysocks,

Thank you for the review! I was able to replicate the button problem but I think it was not related to this PR as after the rebase it is now not happening:
screen shot 2018-04-19 at 18 16 59


Hi @karmatosed, from the UI&UX perspective, I think we can try to use more transparencies or contrasts after the merge and see if we get good results. But for now, do you think this contains the requirements for the merge or you would prefer to have some improvements now?

**Comment by jasmussen** [comment]
> Hi @karmatosed, from the UI&UX perspective, I think we can try to use more transparencies or contrasts after the merge and see if we get good results. But for now, do you think this contains the requirements for the merge or you would prefer to have some improvements now?

Up to Tammie on whether it's ready for merge. But I just pushed a very early work in progress branch, `try/bg-color-friendly`, where I'm doing some work to make the block UI more friendly to colors. You can sort of see an in development progress in a screenshot in https://github.com/WordPress/gutenberg/pull/6237#issue-182410613, — no, there's nothing to see there, just that there's a background color because I'm working on the transparencies there. 

We know we want to make Gutenberg work with a multitude of background colors, even if the default is white.

So the question becomes — is this branch good enough to ship before we address that? Or should we wait for the bg color tweaks to drop first and then rebase?

**Comment by afercia** [comment]
Accessibility note: from a keyboard interaction and regarding the `Alt+F10` shortcut (used to move focus from the editable field to the related toolbar) this improves things! It's actually a good thing to have just _one_ toolbar rendered in a block at a time.

Compared with the previous situation where up to 3 toolbars can be rendered at the same time:

screen shot 2018-04-20 at 08 43 43

For more details see https://github.com/WordPress/gutenberg/pull/6302 which highlights how `focusToolbar()` in `NavigableToolbar` gets all the toolbars. There's room for improvements there, as it's still an issue for example for the Image block caption. Of course, it should be addressed separately.

**Comment by afercia** [comment]
Noticed the "Select parent block" button is empty (no text, no aria-label). Any UI control should be labeled so it has an accessible name assistive technologies can use. It should be changed to a `IconButton` and the text passed as the `label` prop. Seems to work for me, but I'm not sure why the current Dashicon gets passed an `uid` prop.

Also, can this button be added to the toolbar too? Currently, it shows on hover:

screen shot 2018-04-20 at 12 02 51

Adding it to the toolbar, maybe as the first control in the set, would make it usable with a keyboard:

screen shot 2018-04-20 at 12 03 45

**Comment by jorgefilipecosta** [comment]
Hi @afercia, thank you for reviewing this changes from the accessibility point of view. The problems identified are not something new from this PR so I will try to address them in separate PR's.

**Comment by mcsf** [comment]
I think moving forward with this is the only way to tackle all the rough edges and a11y issues. @jorgefilipecosta, is there a way we can help, or does this just need a rebase?

Secondly, if we'd like subsequent PRs to improve this work before a release, it may be best to move this one to 2.9 instead.

**Comment by jasmussen** [comment]
Wanted to note that I'm doing some work in https://github.com/WordPress/gutenberg/pull/6406 that might benefit some of the color/UI issues.

**Comment by jorgefilipecosta** [comment]
Hi @mcsf, the rebase was done. I think this should ready. I'm solving some accessibility issues using other PR's (as they are not exclusive in this PR).
Regarding the design, many improvements are already present and @jasmussen is making things even better in a side PR. I was just waiting for a green light from the UX&UI perspective.

**Review by karmatosed — APPROVED** [review]
I think whilst it's not perfect visually getting this in and refining with the work @jasmussen is doing there, is the best route. As a result approving this and let's note though in merge that this is to be refined, just to set expectations.

**Comment by jasmussen** [comment]
Since this is a pretty complex PR, should we wait for Matías technical thoughts? He's AFK but hopefully back next week. It feels like his comments starting in https://github.com/WordPress/gutenberg/pull/5452#issuecomment-380112310 were sort of skeptical. Not saying because this isn't impressive work, just wondering.

**Comment by gziolo** [comment]
@jorgefilipecosta - can you rebase with the recent changes from `master`?

@mtias - it waits for your feedback.

**Comment by SuperGeniusZeb** [comment]
Should this be added to the Merge Proposal: Editor milestone? It seems like something that would be good for optimum conversion of existing content into Gutenberg block form.

**Comment by SuperGeniusZeb** [comment]
I think the Cover Image block would be made redundant by the existence of a Container block. The Cover Image concept could just be a Reusable block inserted as a template. Think about it. Every setting for the Cover Image block would work for a Container block, and adding nesting to the Cover Image block just makes it into a sort of container block anyway.

See also:
- #4900
- #5452
- #8403

**Pull request closed without merging**
